database/src/main/java/com/firebase/ui/database/paging/FirebaseDataSource.java

```diff
@@ -0,0 +1,182 @@
+package com.firebase.ui.database.paging;
+
+import android.arch.lifecycle.LiveData;
+import android.arch.lifecycle.MutableLiveData;
+import android.arch.paging.DataSource;
+import android.arch.paging.PageKeyedDataSource;
+import android.support.annotation.NonNull;
+import android.support.annotation.RestrictTo;
+
+import com.google.firebase.database.DataSnapshot;
+import com.google.firebase.database.DatabaseError;
+import com.google.firebase.database.Query;
+import com.google.firebase.database.ValueEventListener;
+import com.google.firebase.database.annotations.NotNull;
+
+import java.util.ArrayList;
+
+/**
+ * Data source to power a {@link FirebaseRecyclerPagingAdapter}.
+ *
+ * Note: although loadInitial, loadBefore, and loadAfter are not called on the main thread by the
+ *       paging library, we treat them as if they were so that we can facilitate retry without
+ *       managing our own thread pool or requiring the user to pass us an executor.
+ */
+@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
+public class FirebaseDataSource<T> extends PageKeyedDataSource<String, T> {
```

Comment:
If you look at the Firestore implementation, we have the DataSource deal only in snapshots rather than putting in `Class<T>`.  I think it's simpler that way and also I would like for this code to be as similar to the Firestore code as possible to enable future maintenance.